### PR TITLE
Dev/guest registration

### PR DESCRIPTION
**Pull request by evan-mccarty**
Guest reg page is built, hiding/showing regions in the workflow are WIP. Alex is working on form stuff currently. some tweaks to login page.

### CODE REVIEW

**Comment by coleji** [comment]
@evan-mccarty Can you get rid of the first three files (applicationconfig and the two binary files)?  gitignore them if you need them for your dev environment but they don't need to be committed

**Comment by coleji** [comment]
@evan-mccarty  Put the package-lock file back the way it was (lockfile v2).  Make sure you are using node v14 and npm >= 7.8.0

File: public/waivers/live/ApGuestWaiver.html [hunk 1]
```diff
@@ -0,0 +1,124 @@
+<!DOCTYPE html>
```
Comment (coleji) [inline public/waivers/live/ApGuestWaiver.html:1]: @evan-mccarty Any reason these need to be whole html files iframe'd in?  Seems like they should just be a react component that takes the waiver text as prop, especially since there is so much shared code between these two files.  Also I discussed with Andrew and I think we will at some point have the waiver text come from the backend, although that's not a stopship

**Review by coleji — CHANGES_REQUESTED** [review]

File: public/waivers/live/ApGuestWaiver.html (hunk 1 above)
Comment (evan-mccarty) [reply public/waivers/live/ApGuestWaiver.html:1]: I did this for two reasons. First, it's easier printing the waiver text from the html in the iframe than making a separate print view in TS. Second is a rudimentary version control until we implement a permanent solution.

All the file has is the SVG logo and the waiver text so its pretty small.

If you can think of a more elegant solution,  I'm all ears. (and /or keyboard)

EM

**Comment by evan-mccarty** [comment]
> @evan-mccarty Put the package-lock file back the way it was (lockfile v2). Make sure you are using node v14 and npm >= 7.8.0

I think I have this fixed? Not sure what happened since this came from release/facta

**Comment by evan-mccarty** [comment]
> @evan-mccarty Can you get rid of the first three files (applicationconfig and the two binary files)? gitignore them if you need them for your dev environment but they don't need to be committed

.vs is gone

File: src/containers/ap/ap-registration/ApRequiredInfo.tsx [hunk 2]
```diff
@@ -217,21 +217,6 @@ export default class ApRequiredInfo extends React.Component<Props, State> {
 					options={countries}
 					nullDisplay="- Select -"
 				/>
-				<PhoneTriBox<Form,  PhoneTriBoxProps<Form>>
```
Comment (coleji) [inline src/containers/ap/ap-registration/ApRequiredInfo.tsx]: Did you cut this when you meant to copy?

**Review by coleji — CHANGES_REQUESTED** [review]

File: public/waivers/live/ApGuestWaiver.html (hunk 1 above)
Comment (coleji) [reply public/waivers/live/ApGuestWaiver.html:1]: Oh neat, I didn't see the print button.  Yeah that sounds fine then, probably simple than trying to figure out how to just print a single div

File: src/containers/ap/ap-registration/ApRequiredInfo.tsx (hunk 2 above)
Comment (evan-mccarty) [reply src/containers/ap/ap-registration/ApRequiredInfo.tsx]: good catch, bad Evan

EM

Comment (coleji) [reply src/containers/ap/ap-registration/ApRequiredInfo.tsx]: It's not diffing cleanly, not the end of the world but the easiest way to make it be exactly the way it was before is to run `git show release/facta:path/to/file > path/to/file`.  Same thing happened to package-lock which I already fixed.  Give it a try on this one

**Review by coleji — CHANGES_REQUESTED** [review]

File: src/containers/ap/ApGuestRegister.tsx [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+				{ progressButton }
+				</td>
+				</tr>
+			</tbody></table>
+		</div>);
+
+		const adultWaiverContent = (<div id="adultwaiver">
+		<table width="100%"><tbody>
+				<tr>
+					<iframe title="Waiver of Liability" src="../../../waivers/live/ApGuestWaiver.html" width="100%" height="400px"></iframe>
```
Comment (coleji) [inline src/containers/ap/ApGuestRegister.tsx]: Change this to an absolute path, otherwise if we restructure the paths of the actual pages, it might not be exactly three levels below root anymore.  /public will always be at the URL root.  I'm tempted to say it should be turned into a PathWrapper but for a public thing it's probably fine without

```suggestion
					<iframe title="Waiver of Liability" src="/waivers/live/ApGuestWaiver.html" width="100%" height="400px"></iframe>
```

File: src/containers/ap/ApGuestRegister.tsx [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+				<tr>
+					<td>Agree and Continue</td>
+				</tr>
+			</tbody></table>
+		</div>);
+		const under18WaiverContent = (
+		<div id="under18waiver">
+			<table width="100%"><tbody>
+				<tr>
+					<iframe title="Waiver of Liability" src="../../../waivers/live/Under18GuestWaiver.html" width="100%" height="400px"></iframe>
```
Comment (coleji) [inline src/containers/ap/ApGuestRegister.tsx]: Same as above

File: src/containers/ap/ApGuestRegister.tsx [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+}
+
+type Form = typeof defaultForm
+
+class FormInput extends TextInput<Form> {}
+
+const validate: (state: State) => string[] = state => {
+	const isNotNull = (os: Option<string>) => os.isSome() && os.getOrElse("").length > 0;
+	//const emailValid = (os : Option<string>) => os.isSome() && os.getOrElse("")
+	const validations = [
```
Comment (coleji) [inline src/containers/ap/ApGuestRegister.tsx:82]: My gut reaction is to say, don't bother doing too many validations client-side, most of these things probably really belong server-side (otherwise someone could always just curl the create-person endpoint directly and bypass all this logic in the client).  And once the logic lives on the server, it doesn't also need to live on the client, hence why there aren't a huge number of validations on other pages.  This is not an issue for a closed client like the kiosk where the endpoints are authenticated, but for a web client it is

That said, actually doing these validations server side would be awkward, since I'm assuming you guys won't actually call the server until the final submit on the last page.  We could do a sort of "preflight" api call when they finish the info screen, and have the server save the fields as a proto-person record, and then when they submit the final screen it turns that record into a real person.  But that all seems like overkill

So I think that is all to say, doing these validations on the client and not on the server opens us up to that risk of someone posting to the api server directly without using the client and skipping the validations, but in this case I think the cost of doing it right outweighs the benefit, so let's leave this as is.

Validations that the server runs on the member page which you should replicate here are

- DOB is a valid date (e.g. 2/31/1990 should throw an error)
- Email is a valid email (I use the regex  `/^[A-Z0-9._%-]+@[A-Z0-9._%-]+\.[A-Z]{2,4}$/i`  )
- Phone number is valid (matches  `/^[0-9]{10}.*$/`  ie exactly 10 digits, with the .* for any extension )

https://github.com/community-boating/cbidb-api/blob/master/app/org/sailcbi/APIServer/Api/Endpoints/Member/APRequiredInfo.scala

**Review by coleji — CHANGES_REQUESTED** [review]

File: ie/ie.html [hunk 6]
```diff
@@ -1,6 +1,521 @@
 <html class="no-js" lang="en-US">
 
 <head>
+<<<<<<< HEAD
```
Comment (coleji) [inline ie/ie.html]: looks like this wasn't merged properly

File: ini/config.ini.template [hunk 7]
```diff
@@ -1,13 +0,0 @@
-stripeKey=ask_jon
```
Comment (coleji) [inline ini/config.ini.template]: Please put this file back

File: src/containers/ap/ApGuestRegister.tsx [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+					});
+				}
+			break;
+		}
+		return Promise.resolve("Success");
+	}
+	
+	render() {
+		const thisYear = Number(moment().format("YYYY"))
+		const years = range(thisYear-120, thisYear-15)
```
Comment (coleji) [inline src/containers/ap/ApGuestRegister.tsx]: Go right up to current year, I don't think its unheard of to bring very small children out as guests

File: src/containers/ap/ApGuestRegister.tsx [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+		);
+
+		const formContent = (
+		<div id="guestinfo">
+			Guests must self register. Members please direct your guests to this page to register themselves. Under 18 guests must be registered by their parent or guardian.
+			<br />
+			At the end of registration you will be able to print or save your ticket, you'll also receive an email ticket.
+			<br />
+				<table id="info" style={{ width: "100%"}}><tbody>
+					<th>Guest Information</th>
```
Comment (coleji) [inline src/containers/ap/ApGuestRegister.tsx]: th should appear inside a tr

File: src/containers/ap/ApGuestRegister.tsx [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+				id="email"
+				label="Email"
+				isPassword={false}
+				isRequired
+				value={self.state.formData.email}
+				updateAction={updateState}
+				/>
+				</tbody></table>
+				<table id="ecInfo" style={{ width: "100%",}}><tbody>
+				<th>Emergency Contact Information</th>
```
Comment (coleji) [inline src/containers/ap/ApGuestRegister.tsx]: same as above

File: src/containers/ap/ApGuestRegister.tsx [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+				<tr><td />This should be someone close to you who is not going on the water with you.</tr>
```
Comment (coleji) [inline src/containers/ap/ApGuestRegister.tsx]: Text should be inside the td

File: src/containers/ap/ApGuestRegister.tsx [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+				firstValue={self.state.formData.ecPhoneFirst}
+				secondValue={self.state.formData.ecPhoneSecond}
+				thirdValue={self.state.formData.ecPhoneThird}
+				extValue={self.state.formData.ecPhoneExt}
+				typeValue={self.state.formData.ecPhoneType}
+				updateAction={updateState}
+				isRequired={true}
+				/>
+				<tr>
+				{ progressButton }
```
Comment (coleji) [inline src/containers/ap/ApGuestRegister.tsx]: button should be inside a td (if its going to be inside a tr)

File: src/containers/ap/ApGuestRegister.tsx [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+				<tr>
+					<td>{ agreeCheckbox } { progressButton }</td>
+				</tr>
+			</tbody></table>
+		</div>);
+		const under18WaiverContent = (
+		<div id="under18waiver">
+			<table width="100%"><tbody>
+				<tr>
+					<iframe title="Waiver of Liability" src="/waivers/live/Under18GuestWaiver.html" width="100%" height="400px"></iframe>
```
Comment (coleji) [inline src/containers/ap/ApGuestRegister.tsx]: iframe should be in a td

**Review by coleji — CHANGES_REQUESTED** [review]
Lots of semantic issues.  Check the browser console for errors about dom nodes that should not be directly inside other dom nodes (e.g. text directly in a tr, without being in a td)

Package-lock got deleted again, and config.ini.template.  ie.html still has merge conflict artifacts in it.  You should proofread the diff to make sure there are no unexpected changes.

The validations I asked for last time don't seem to have been implemented yet (eg email and phone regex check)

Otherwise the content/functionality look good.  I will probably create a new api endpoint that combines create-person and create-card into one, and has some level of flood prevention/idempotency for public use rather than just unlock the kiosk endpoint.  If/when I do that, I'll make that change; you can leave it as is

File: src/containers/ap/ApGuestRegister.tsx (hunk 9 above)
Comment (evan-mccarty) [reply src/containers/ap/ApGuestRegister.tsx]: which line are you referencing 264 or 266

File: src/containers/ap/ApGuestRegister.tsx (hunk 10 above)
Comment (evan-mccarty) [reply src/containers/ap/ApGuestRegister.tsx]: 320?

File: src/containers/ap/ApGuestRegister.tsx (hunk 9 above)
Comment (coleji) [reply src/containers/ap/ApGuestRegister.tsx]: 266/267

File: src/containers/ap/ApGuestRegister.tsx (hunk 10 above)
Comment (coleji) [reply src/containers/ap/ApGuestRegister.tsx]: yep

File: src/containers/ap/ApGuestRegister.tsx (hunk 12 above)
Comment (evan-mccarty) [reply src/containers/ap/ApGuestRegister.tsx]: would you prefer it attached to the last form input? I have no strong preference.

Comment (coleji) [reply src/containers/ap/ApGuestRegister.tsx]: You can leave it there and just put `<td>`'s around it.  I don't care where it actually is, all I'm saying is you can't put things that aren't `<td>` or `<th>` directly under a `<tr>` (and you can't put `<th>` or `<td>` anywhere else)

File: src/containers/ap/ApGuestRegister.tsx (hunk 9 above)
Comment (evan-mccarty) [reply src/containers/ap/ApGuestRegister.tsx:267]: nevermind i read that wrong: however i believe th does not require a containing td

File: ie/ie.html (hunk 6 above)
Comment (evan-mccarty) [reply ie/ie.html]: looks like you're right!

Comment (coleji) [reply ie/ie.html]: You resolved this conversation but the file is still busted

File: ini/config.ini.template (hunk 7 above)
Comment (coleji) [reply ini/config.ini.template]: You resolved this conversation but the file is still deleted.  Also package-lock.json is still deleted (GH won't let me comment on that file)

File: maintenance/maintenance.html [hunk 14]
```diff
@@ -192,7 +192,7 @@
 				<div class="page-title">
 					<div class="ratio"></div>
 					<div class="cover"
-						style="background-image: url(/facta/assets/images/common-hero.jpg);"></div>
+						style="background-image: url(&quot;/facta/assets/images/IMG_1495-scaled.jpeg&quot;);"></div>
```
Comment (coleji) [inline maintenance/maintenance.html]: What is the purpose of this change? This file doesn't exist

File: src/components/PhoneTriBox.tsx [hunk 15]
```diff
@@ -113,7 +112,7 @@ export default class PhoneTriBox<U, T extends PhoneTriBoxProps<U>> extends React
 			<Select<U>
 				id={self.props.typeID}
 				label="Type"
-				isRequired={self.props.isRequired}
+				//isRequired={self.props.isRequired}
```
Comment (coleji) [inline src/components/PhoneTriBox.tsx]: This will affect all PhoneTriBox elements, not just yours.  Please put this back the way it was, and mark your own as not required if that's what you want

**Review by coleji — CHANGES_REQUESTED** [review]

File: maintenance/maintenance.html (hunk 14 above)
Comment (evan-mccarty) [reply maintenance/maintenance.html]: I will ask Alex but I dont think that was our change..

File: src/components/PhoneTriBox.tsx (hunk 15 above)
Comment (evan-mccarty) [reply src/components/PhoneTriBox.tsx]: fixed

File: maintenance/maintenance.html (hunk 14 above)
Comment (coleji) [reply maintenance/maintenance.html]: You made that change last weekend
https://github.com/community-boating/cbidb-public-web/commit/dc040b193d946caf4a208982d6ed2d966d4fb167

Comment (evan-mccarty) [reply maintenance/maintenance.html]: mustve been a copy and paste error, ill revert it

**Review by coleji — APPROVED** [review]
